### PR TITLE
Multi-level Nested Sort with Filters

### PR DESCRIPTION
**Pull request by mattweber**
Allow multiple levels of nested sorting where each level can have it's own filter. This allows us to have complex nested sort criteria just like we have with queries and aggregations.   This is backwards compatible with previous single-level nested sort.

@martijnvg @jpountz Can you please review?

### CODE REVIEW

File: docs/reference/search/request/sort.asciidoc
Comment (martijnvg) [inline docs/reference/search/request/sort.asciidoc]: The migration should document that `nested_path` and `nested_filter` are deprecated.

File: core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java
Comment (martijnvg) [inline core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java]: I think we should add some deprecation logging here too for the `nested_filter` and `nested_path` parameters.

```
private static final DeprecationLogger DEPRECATION_LOGGER = new DeprecationLogger(Loggers.getLogger(ParseField.class));
...
PARSER.declareString((fieldSortBuilder, nestedPath) -> {
            DEPRECATION_LOGGER.deprecated("[nested_path] has been deprecated in favor of the [nested] parameter");
            fieldSortBuilder.setNestedPath(nestedPath);
        } , NESTED_PATH_FIELD);
...
PARSER.declareObject(FieldSortBuilder::setNestedFilter, (p, c) -> {
            DEPRECATION_LOGGER.deprecated("[nested_filter] has been deprecated in favour for the [nested] parameter");
            return SortBuilder.parseNestedFilter(p);}, NESTED_FILTER_FIELD);
...
```

**Review by martijnvg — COMMENTED** [review]
@mattweber This looks great and I like the new syntax for nested sorting! I left a few comments.

**Comment by martijnvg** [comment]
test this please

File: core/src/main/java/org/elasticsearch/search/sort/SortBuilder.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/sort/SortBuilder.java]: Even though it is not strictly a sort builder, can you rename it NestedSortBuilder for consistency and move it to its own class?

**Comment by mattweber** [comment]
@martijnvg Thanks for the review!  Thats a bummer about waiting for 6.1, I was hoping it could make a 5.x release since it is backward compatible.  I have addressed your comments, please let me know if you would like any other changes.

File: core/src/main/java/org/elasticsearch/search/sort/SortBuilder.java
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/search/sort/SortBuilder.java]: should we requite the path in the constructor since it is mandatory?

**Review by jpountz — COMMENTED** [review]
Even thoug I think nested is often abused, I don't like the fact that queries and aggregations allow for multiple levels of nesting but sorting doesn't, so +1 to expose a way to make it work.

**Comment by mattweber** [comment]
Thanks @jpountz, I addressed your comments.  How does this look?

**Comment by martijnvg** [comment]
retest this please

**Comment by martijnvg** [comment]
@mattweber Thanks for fixing the comments. The PR looks good! Only the build failed with checkstyle violations. Can you fix this?

**Review by jpountz — APPROVED** [review]

**Comment by mattweber** [comment]
@martijnvg @jpountz Thank you for reviewing everything.  Just pushed checkstyle and test fixes.  I think this will do it.

**Comment by martijnvg** [comment]
retest this please

**Comment by martijnvg** [comment]
The mixed cluster test failures that occurred in the build this PR triggered are related to the fact that the changes aren't yet in the 6.x branch. This causes serialization errors when a 7.0.0-alph1 node reads the shard level search request from a 6.1.0-SNAPSHOT node. So I'll merge this change both to master and 6.x branches at the same time to minimize the regular builds from running into this failure.

**Comment by jasontedor** [comment]
The versions and skips should be set so that they cause no failures, backported with version updates, then another commit to master moving the version to the 6.x version they are compatible with.

**Comment by jasontedor** [comment]
Here's another PR where I [expand](https://github.com/elastic/elasticsearch/pull/26166#issuecomment-322140800) on this strategy; this is build failure minimizing strategy, not merging as quickly as possible. Instead, we should take our time to get a green build on master, then backport (with the version bumps) and getting a green build again, and then a final commit to get the versions right still maintaining green status.

**Comment by martijnvg** [comment]
@jasontedor Good point. I'll follow the same steps mentioned in that PR.

**Comment by martijnvg** [comment]
I merged squashed the change into the master branch via 6377afa6c376e4c7d0d477e1d4e5f9dbd222a3c4
and set the version to 7.0.0-alph1 via 472a5dd56b4360037dc7ce7c2837c929b3c3fd25

I'll backport this change to 6.x branch later today unless a build failure related to this change occurs.

@mattweber Thanks for this contribution to nested sorting!

**Comment by martijnvg** [comment]
I just realized that as part of the git merge (git merge --squash) that I became the author of this change :( 
Not sure what would be best to restore this mistake. (I can only think of reverting the commit and then commit it again with the @mattweber as author)

**Comment by jasontedor** [comment]
@martijnvg For future reference, you can push commits to PRs so you could have pushed the version-changing commit to this PR, then merged from the GitHub UI (preserving @mattweber as the author). As far as correcting, I think that you can:
 - revert the commits that you pushed to master
 - reopen this PR
 - push the version-changing commit to this PR
 - merge this PR from the GitHub UI

**Comment by martijnvg** [comment]
@mattweber Can you change the 6.1.0 version check to 7.0.0-alpha1? (as it turns out I can't make this change in your pr) After that I'll hit the squash and merge button and do they backporting to 6.x branch.

**Comment by jpountz** [comment]
@martijnvg I'm wondering whether that would work with the online editor, eg. https://github.com/mattweber/elasticsearch/edit/multi_nested_sort/core/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java?pr=/elastic/elasticsearch/pull/26395

**Comment by martijnvg** [comment]
retest this please

**Comment by martijnvg** [comment]
@jpountz Your idea worked :) 

I'll squash and merge once the build is green.

**Comment by mattweber** [comment]
Wow lots going on.  Thanks @martijnvg @jpountz @jasontedor!  Let me know if you need me to do anything.

**Comment by martijnvg** [comment]
backported to 6.x branch via: 2425da1ccd65846c0fb2721cfe6c2e21d2f77e34 and 42c774583a52fc3856615747f14c3750db43d560

Changed the version check in master via ea3fa768f95a0645684e153d41563bcd9117fd1b

**Pull request merged**
